File: Documentation/5.4/Samples/python/ClientApi/Session/Querying/HowToPerformFacetedSearch.py
import enum
from abc import ABC
from typing import Dict, Union, Set, Optional, List, Callable, TypeVar

from ravendb import (
    Facet,
    RangeFacet,
    FacetAggregationField,
    FacetBase,
    AggregationDocumentQuery,
    FacetBuilder,
    FacetOperations,
    RangeBuilder,
    IndexDefinition,
    PutIndexesOperation,
)
from ravendb.documents.queries.facets.definitions import FacetSetup
from ravendb.documents.queries.facets.misc import FacetOptions, FacetTermSortMode, FacetAggregation
from ravendb.primitives import constants

from cameras_helper import get_cameras, Camera
from examples_base import ExampleBase

_T = TypeVar("_T")


class HowToPerformFacetedSearch(ExampleBase):
    def setUp(self):
        super().setUp()
        with self.embedded_server.get_document_store("FacetedSearch") as store:
            with store.bulk_insert() as bulk_insert:
                for camera in get_cameras(100):
                    bulk_insert.store(camera)

            index_definition = IndexDefinition()
            index_definition.name = "Camera/Costs"
            index_definition.maps = [
                "from camera in docs select new { camera.manufacturer, camera.model, camera.cost, camera.date_of_listing, camera.megapixels } "
            ]
            store.maintenance.send(PutIndexesOperation(index_definition))

    def test_how_to_perform_faceted_search(self):
        with self.embedded_server.get_document_store("FacetedSearch") as store:
            with store.open_session() as session:
                # region facet_2_1
                facet_options = FacetOptions.default_options()
                facet_options.term_sort_mode = FacetTermSortMode.COUNT_DESC
                facet_options.start = 0

                facet1 = Facet("manufacturer")
                facet1.options = facet_options

                facet2 = RangeFacet()
                facet2.ranges = [
                    "cost < 200",
                    "cost between 200 and 400",
                    "cost between 400 and 600",
                    "cost between 600 and 800",
                    "cost >= 800",
                ]
                facet2.aggregations = {FacetAggregation.AVERAGE: {FacetAggregationField("cost")}}

                facet3 = RangeFacet()
                facet3.ranges = [
                    "megapixels < 3",
                    "megapixels between 3 and 7",
                    "megapixels between 7 and 10",
                    "megapixels >= 10",
                ]

                facets = (
                    session.query_index("Camera/Costs", Camera)
                    .aggregate_by(facet1)
                    .and_aggregate_by(facet2)
                    .and_aggregate_by(facet3)
                    .execute()
                )
                # endregion

            with store.open_session() as session:
                # region facet_3_1
                options = FacetOptions()
                options.start = 0
                options.term_sort_mode = FacetTermSortMode.COUNT_DESC

                cost_builder = RangeBuilder.for_path("cost")
                megapixels_builder = RangeBuilder.for_path("megapixels")

                facet_result = (
                    session.query_index("Camera/Costs", Camera)
                    .aggregate_by(lambda builder: builder.by_field("manufacturer").with_options(options))
                    .and_aggregate_by(
                        lambda builder: builder.by_ranges(
                            cost_builder.is_less_than(200),
                            cost_builder.is_greater_than_or_equal_to(200).is_less_than(400),
                            cost_builder.is_greater_than_or_equal_to(400).is_less_than(600),
                            cost_builder.is_greater_than_or_equal_to(600).is_less_than(800),
                            cost_builder.is_greater_than_or_equal_to(800),
                        ).average_on("cost")
                    )
                    .and_aggregate_by(
                        lambda builder: builder.by_ranges(
                            megapixels_builder.is_less_than(3),
                            megapixels_builder.is_greater_than_or_equal_to(3).is_less_than(7),
                            megapixels_builder.is_greater_than_or_equal_to(7).is_less_than(10),
                            megapixels_builder.is_greater_than_or_equal_to(10),
                        )
                    )
                ).execute()
                # endregion

            with store.open_session() as session:
                # region facet_4_1
                facet_setup = FacetSetup()

                facet_manufacturer = Facet()
                facet_manufacturer.field_name = "manufacturer"
                facet_setup.facets = [facet_manufacturer]

                camera_facet = RangeFacet()
                camera_facet.ranges = [
                    "cost < 200",
                    "cost between 200 and 400",
                    "cost between 400 and 600",
                    "cost between 600 and 800",
                    "cost >= 800",
                ]

                megapixels_facet = RangeFacet()
                megapixels_facet.ranges = [
                    "megapixels < 3",
                    "megapixels between 3 and 7",
                    "megapixels between 7 and 10",
                    "megapixels >= 10",
                ]

                facet_setup.range_facets = [camera_facet, megapixels_facet]

                session.store(facet_setup, "facets/CameraFacets")
                session.save_changes()

                facets = session.query_index("Camera/Costs", Camera).aggregate_using("facets/CameraFacets").execute()
                # endregion

    # region facet_7_3
    class FacetBase(ABC):
        def __init__(self):
            self.display_field_name: Union[None, str] = None
            self.options: Union[None, FacetOptions] = None
            self.aggregations: Dict[FacetAggregation, Set[FacetAggregationField]] = {}

    class Facet(FacetBase):
        def __init__(self, field_name: str = None):
            super().__init__()
            self.field_name = field_name

    # endregion

    # region facet_7_4
    class RangeFacet(FacetBase):
        def __init__(self, parent: Optional[FacetBase] = None):
            super().__init__()
            self.ranges: List[str] = []

    # endregion

    # region facet_7_5
    class FacetAggregation(enum.Enum):
        NONE = "None"
        MAX = "Max"
        MIN = "Min"
        AVERAGE = "Average"
        SUM = "Sum"

    # endregion

    class Foo:
        # region facet_1
        def aggregate_by(
            self, builder_or_facet: Union[Callable[[FacetBuilder], None], FacetBase]
        ) -> AggregationDocumentQuery[_T]: ...

        def aggregate_by_facets(self, facets: List[FacetBase]) -> AggregationDocumentQuery[_T]: ...

        def aggregate_using(self, facet_setup_document_id: str) -> AggregationDocumentQuery[_T]: ...

        # endregion

        # region facet_7_1
        def by_ranges(self, range_: RangeBuilder, *ranges: RangeBuilder) -> FacetOperations[_T]: ...

        def by_field(self, field_name: str) -> FacetOperations[_T]: ...

        def with_display_name(self, display_name: str) -> FacetOperations[_T]: ...

        def with_options(self, options: FacetOptions) -> FacetOperations[_T]: ...

        def sum_on(self, path: str, display_name: Optional[str] = None) -> FacetOperations[_T]: ...

        def min_on(self, path: str, display_name: Optional[str] = None) -> FacetOperations[_T]: ...

        def max_on(self, path: str, display_name: Optional[str] = None) -> FacetOperations[_T]: ...

        def average_on(self, path: str, display_name: Optional[str] = None) -> FacetOperations[_T]: ...

        # endregion

    class Foo1:
        # region facet_7_2
        def __init__(self):
            self.page_size: int = constants.int_max
            self.start: Union[None, int] = None
            self.term_sort_mode: FacetTermSortMode = FacetTermSortMode.VALUE_ASC
            self.include_remaining_terms: bool = False

        # endregion

File: Documentation/5.4/Samples/php/ClientApi/session/Querying/HowToPerformFacetedSearch.php
<?php

use RavenDB\Documents\Queries\Facets\AggregationArray;
use RavenDB\Documents\Queries\Facets\AggregationDocumentQueryInterface;
use RavenDB\Documents\Queries\Facets\FacetOperationsInterface;
use RavenDB\Documents\Queries\Facets\FacetOptions;
use RavenDB\Documents\Queries\Facets\FacetResult;
use RavenDB\Documents\Queries\Facets\FacetSetup;
use RavenDB\Documents\Queries\Facets\FacetTermSortMode;
use RavenDB\Documents\Queries\Facets\RangeBuilder;
use RavenDB\Documents\Queries\Query;
use RavenDB\Documents\Session\Tokens\FacetToken;
use RavenDB\Samples\Infrastructure\DocumentStoreHolder;
use RavenDB\Type\StringArray;


//region facet_7_3
abstract class FacetBase
{
    /** @SerializedName("DisplayFieldName") */
    private ?string $displayFieldName = null;

    /** @SerializedName("Options") */
    private ?FacetOptions $options = null;

    /** @SerializedName("Aggregations") */
    private ?AggregationArray $aggregations = null;

    public function __construct()
    {
        $this->aggregations = new AggregationArray();
    }

    public function getDisplayFieldName(): ?string
    {
        return $this->displayFieldName;
    }

    public function setDisplayFieldName(?string $displayFieldName): void
    {
        $this->displayFieldName = $displayFieldName;
    }

    public function getOptions(): ?FacetOptions
    {
        return $this->options;
    }

    public function setOptions(?FacetOptions $options): void
    {
        $this->options = $options;
    }

    public function getAggregations(): ?AggregationArray
    {
        return $this->aggregations;
    }

    public function setAggregations(null|array|StringArray $aggregations): void
    {
        if (is_array($aggregations)) {
            $aggregations = AggregationArray::fromArray($aggregations);
        }
        $this->aggregations = $aggregations;
    }

    public abstract function toFacetToken(Closure $addQueryParameter): FacetToken;
}

class Facet extends FacetBase
{
    /** @SerializedName("FieldName") */
    private string $fieldName;

    public function getFieldName(): string
    {
        return $this->fieldName;
    }

    public function setFieldName(string $fieldName): void
    {
        $this->fieldName = $fieldName;
    }

    public function toFacetToken(Closure $addQueryParameter): FacetToken
    {
        return FacetToken::create($this, $addQueryParameter);
    }
}
//endregion


//region facet_7_4
class RangeFacet extends FacetBase
{
    private ?FacetBase $parent = null;

    /**
     * @SerializedName("Ranges")
     */
    private StringArray $ranges;

    public function __construct(?FacetBase $parent = null)
    {
        parent::__construct();

        $this->parent = $parent;
        $this->ranges = new StringArray();
    }

    public function getRanges(): StringArray
    {
        return $this->ranges;
    }

    /**
     * @param array|StringArray $ranges
     */
    public function setRanges(array|StringArray $ranges): void
    {
        if (is_array($ranges)) {
            $ranges = StringArray::fromArray($ranges);
        }
        $this->ranges = $ranges;
    }

    public function toFacetToken($addQueryParameter): FacetToken
    {
        if ($this->parent != null) {
            return $this->parent->toFacetToken($addQueryParameter);
        }

        return FacetToken::create($this, $addQueryParameter);
    }
}
//endregion

//region facet_7_5
class FacetAggregation
{
    public const NONE = "NONE";
    public const MAX  = "MAX";
    public const MIN = "MIN";
    public const AVERAGE = "AVERAGE";
    public const SUM = "SUM";
}
//endregion

interface FooInterface
{
    //region facet_1
    /**
     * Usage
     *   - aggregateBy(Callable $builder);
     *   - aggregateBy(FacetBase $facet);
     *   - aggregateBy(Facet... $facets);
     *
     * @param Callable|FacetBase $builderOrFacets
     *
     * @return AggregationDocumentQueryInterface
     */
    public function aggregateBy(...$builderOrFacets): AggregationDocumentQueryInterface;

    public function aggregateUsing(?string $facetSetupDocumentId): AggregationDocumentQueryInterface;
    //endregion

    //region facet_7_1
    public function byRanges(?RangeBuilder $range, ?RangeBuilder ...$ranges): FacetOperationsInterface;

    public function byField(string $fieldName): FacetOperationsInterface;
    public function withDisplayName(string $displayName): FacetOperationsInterface;

    public function withOptions(FacetOptions $options): FacetOperationsInterface;

    public function sumOn(string $path, ?string $displayName = null): FacetOperationsInterface;
    public function minOn(string $path, ?string $displayName = null): FacetOperationsInterface;
    public function maxOn(string $path, ?string $displayName = null): FacetOperationsInterface;
    public function averageOn(string $path, ?string $displayName = null): FacetOperationsInterface;
    //endregion
}

class Foo1 {
    //region facet_7_2
    private FacetTermSortMode $termSortMode;
    private bool $includeRemainingTerms = false;
    private int $start = 0;
    private int $pageSize = 0;

    private const INT_32_MAX = 2147483647;

    public function __construct() {
        $this->pageSize = self::INT_32_MAX;
        $this->termSortMode = FacetTermSortMode::valueAsc();
    }

    //getters and setters
    //endregion
}

class HowToPerformFacetedSearch
{
    public function samples(): void
    {
        $store = DocumentStoreHolder::getStore();
        try {

            $session = $store->openSession();
            try {
                //region facet_2_1
                $facetOptions = new FacetOptions();
                $facetOptions->setTermSortMode(FacetTermSortMode::countDesc());

                $facet1 = new Facet();
                $facet1->setFieldName("manufacturer");
                $facet1->setOptions($facetOptions);

                $facet2 = new RangeFacet();
                $facet2->setRanges([
                    "cost < 200",
                    "cost between 200 and 400",
                    "cost between 400 and 600",
                    "cost between 600 and 800",
                    "cost >= 800"
                ]);
                $facet2->setAggregations([FacetAggregation::AVERAGE, "cost"]);

                $facet3 = new RangeFacet();
                $facet3->setRanges([
                    "megapixels < 3",
                    "megapixels between 3 and 7",
                    "megapixels between 7 and 10",
                    "megapixels >= 10"
                ]);

                /** @var array<FacetResult> $facets */
                $facets = $session
                    ->query(Camera::class, Query::index("Camera/Costs"))
                    ->aggregateBy($facet1)
                    ->andAggregateBy($facet2)
                    ->andAggregateBy($facet3)
                    ->execute();
                //endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                //region facet_3_1
                $options = new FacetOptions();
                $options->setTermSortMode(FacetTermSortMode::countDesc());

                $costBuilder = RangeBuilder::forPath("cost");
                $megapixelsBuilder = RangeBuilder::forPath("megapixels");

                /** @var array<FacetResult> $facetResult */
                $facetResult = $session
                    ->query(Camera::class, Query::index("Camera/Costs"))
                    ->aggregateBy(function($builder) use ($options) {
                        $builder
                            ->byField("manufacturer")
                            ->withOptions($options);
                    })
                    ->andAggregateBy(function($builder) use ($costBuilder) {
                        $builder
                            ->byRanges(
                                $costBuilder->isLessThan(200),
                                $costBuilder->isGreaterThanOrEqualTo(200)->isLessThan(400),
                                $costBuilder->isGreaterThanOrEqualTo(400)->isLessThan(600),
                                $costBuilder->isGreaterThanOrEqualTo(600)->isLessThan(800),
                                $costBuilder->isGreaterThanOrEqualTo(800)
                             )
                            ->averageOn("cost");
                    })
                    ->andAggregateBy(function($builder) use ($megapixelsBuilder) {
                        $builder
                            ->byRanges(
                                $megapixelsBuilder->isLessThan(3),
                                $megapixelsBuilder->isGreaterThanOrEqualTo(3)->isLessThan(7),
                                $megapixelsBuilder->isGreaterThanOrEqualTo(7)->isLessThan(10),
                                $megapixelsBuilder->isGreaterThanOrEqualTo(10)
                            );
                    })
                    ->execute();
                //endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                //region facet_4_1
                $facetSetup = new FacetSetup();

                $facetManufacturer = new Facet();
                $facetManufacturer->setFieldName("manufacturer");
                $facetSetup->setFacets([$facetManufacturer]);

                $cameraFacet = new RangeFacet();
                $cameraFacet->setRanges([
                    "cost < 200",
                    "cost between 200 and 400",
                    "cost between 400 and 600",
                    "cost between 600 and 800",
                    "cost >= 800"
                ]);

                $megapixelsFacet = new RangeFacet();
                $megapixelsFacet->setRanges([
                    "megapixels < 3",
                    "megapixels between 3 and 7",
                    "megapixels between 7 and 10",
                    "megapixels >= 10"
                ]);

                $facetSetup->setRangeFacets([$cameraFacet, $megapixelsFacet]);

                $session->store($facetSetup, "facets/CameraFacets");
                $session->saveChanges();

                /** @var array<FacetResult> $facets */
                $facets = $session
                    ->query(Camera::class, Query::index("Camera/Costs"))
                    ->aggregateUsing("facets/CameraFacets")
                    ->execute();
                //endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}

class Camera
{

}